From 853d393a441f9df7542a70218801d5a2336ade18 Mon Sep 17 00:00:00 2001
From: Christophe Kerello <christophe.kerello@foss.st.com>
Date: Wed, 16 Jul 2025 15:50:05 +0200
Subject: [PATCH 917/954] perf(st-spi): improve indirect read/write throughput

Improve indirect read/write throughput by reading/writing 4-bytes per
4-bytes instead of reading/writing byte per byte.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 209b48da6353138423123673933fb029b81070fa perf(st-spi): improve indirect read/write throughput
73a0ca851ab48986e8fc1f7006879de3f53b9894 perf(st-spi): improve indirect read/write throughput]
Change-Id: Ief34af60c153d89bc17062c114a393c33a346aa4
Signed-off-by: Christophe Kerello <christophe.kerello@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/479644
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Christophe KERELLO <christophe.kerello@st.com>
---
 drivers/st/spi/stm32_ospi.c | 72 ++++++++++++++++++++++++++++++++-----
 1 file changed, 64 insertions(+), 8 deletions(-)

diff --git a/drivers/st/spi/stm32_ospi.c b/drivers/st/spi/stm32_ospi.c
index b7d731244b..748ebf3597 100644
--- a/drivers/st/spi/stm32_ospi.c
+++ b/drivers/st/spi/stm32_ospi.c
@@ -47,6 +47,8 @@
 /* OCTOSPI control register */
 #define _OSPI_CR_EN		BIT(0)
 #define _OSPI_CR_ABORT		BIT(1)
+#define _OSPI_CR_FTHRES		GENMASK_32(13, 8)
+#define _OSPI_CR_FTHRES_SHIFT	8U
 #define _OSPI_CR_CSSEL		BIT(24)
 #define _OSPI_CR_FMODE		GENMASK_32(29, 28)
 #define _OSPI_CR_FMODE_SHIFT	28U
@@ -205,20 +207,52 @@ static int stm32_ospi_wait_cmd(void)
 	return ret;
 }
 
-static void stm32_ospi_read_fifo(uint8_t *val, uintptr_t addr)
+static int stm32_ospi_read_fifo(void *val, uintptr_t addr, uint64_t len)
 {
-	*val = mmio_read_8(addr);
+	int ret = 0;
+
+	switch (len) {
+	case sizeof(uint32_t):
+		*((uint32_t *)val) = mmio_read_32(addr);
+		break;
+	case sizeof(uint8_t):
+		*((uint8_t *)val) = mmio_read_8(addr);
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	};
+
+	return ret;
 }
 
-static void stm32_ospi_write_fifo(uint8_t *val, uintptr_t addr)
+static int stm32_ospi_write_fifo(void *val, uintptr_t addr, uint64_t len)
 {
-	mmio_write_8(addr, *val);
+	int ret = 0;
+
+	switch (len) {
+	case sizeof(uint32_t):
+		mmio_write_32(addr, *((uint32_t *)val));
+		break;
+	case sizeof(uint8_t):
+		mmio_write_8(addr, *((uint8_t *)val));
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	};
+
+	return ret;
 }
 
 static int stm32_ospi_poll(uint8_t *buf, uint32_t nbytes, bool read)
 {
-	void (*fifo)(uint8_t *val, uintptr_t addr);
-	uint32_t len;
+	int (*fifo)(void *val, uintptr_t addr, uint64_t len);
+	uint8_t *buff = buf;
+	uint64_t length = nbytes;
+	uint32_t tmp;
+	uint64_t i = 0U;
+	int ret;
 
 	if (read) {
 		fifo = stm32_ospi_read_fifo;
@@ -226,7 +260,7 @@ static int stm32_ospi_poll(uint8_t *buf, uint32_t nbytes, bool read)
 		fifo = stm32_ospi_write_fifo;
 	}
 
-	for (len = nbytes; len != 0U; len--) {
+	while (length != 0U) {
 		uint64_t timeout = timeout_init_us(_OSPI_FIFO_TIMEOUT_US);
 
 		while ((mmio_read_32(ospi_base() + _OSPI_SR) &
@@ -240,7 +274,27 @@ static int stm32_ospi_poll(uint8_t *buf, uint32_t nbytes, bool read)
 			}
 		}
 
-		fifo(buf++, ospi_base() + _OSPI_DR);
+		if (length >= sizeof(uint32_t)) {
+			ret = fifo(&tmp, ospi_base() + _OSPI_DR,
+				   sizeof(uint32_t));
+			if (ret != 0) {
+				return ret;
+			}
+
+			(void)memcpy((void *)&buff[i], (const void *)&tmp,
+				     sizeof(uint32_t));
+			length -= sizeof(uint32_t);
+			i += sizeof(uint32_t);
+		} else {
+			ret = fifo(&buff[i], ospi_base() + _OSPI_DR,
+				   sizeof(uint8_t));
+			if (ret != 0) {
+				return ret;
+			}
+
+			length--;
+			i++;
+		}
 	}
 
 	return 0;
@@ -1334,6 +1388,8 @@ int stm32_ospi_init(void)
 	}
 
 	mmio_write_32(ospi_base() + _OSPI_DCR1, _OSPI_DCR1_DEVSIZE);
+	mmio_clrsetbits_32(ospi_base() + _OSPI_CR, _OSPI_CR_FTHRES,
+			   3U << _OSPI_CR_FTHRES_SHIFT);
 
 #if STM32MP_HYPERFLASH
 	/* Init Hyperflash framework */
-- 
2.53.0

